internal/rbd/controllerserver.go

```diff
@@ -622,49 +731,81 @@ func (cs *ControllerServer) validateSnapshotReq(ctx context.Context, req *csi.Cr
 	return nil
 }
 
-func (cs *ControllerServer) doSnapshot(ctx context.Context, rbdSnap *rbdSnapshot, cr *util.Credentials) (err error) {
-	err = createSnapshot(ctx, rbdSnap, cr)
-	// If snap creation fails, even due to snapname already used, fail, next attempt will get a new
-	// uuid for use as the snap name
+func (cs *ControllerServer) doSnapshot(ctx context.Context, parentVol *rbdVolume, rbdSnap *rbdSnapshot, cr *util.Credentials) (bool, *rbdVolume, error) {
+	// generate cloned volume details from snapshot
+	cloneRbd := generateVolFromSnap(rbdSnap)
+	defer cloneRbd.Destroy()
+	// add image feature for cloneRbd
+	f := []string{librbd.FeatureNameLayering, librbd.FeatureNameDeepFlatten}
+	cloneRbd.imageFeatureSet = librbd.FeatureSetFromNames(f)
+	ready := false
+
+	err := cloneRbd.Connect(cr)
+	if err != nil {
+		return ready, cloneRbd, err
+	}
+
+	err = createRBDClone(ctx, parentVol, cloneRbd, rbdSnap, cr)
 	if err != nil {
 		klog.Errorf(util.Log(ctx, "failed to create snapshot: %v"), err)
-		return status.Error(codes.Internal, err.Error())
+		return ready, cloneRbd, status.Error(codes.Internal, err.Error())
 	}
+
 	defer func() {
 		if err != nil {
-			errDefer := deleteSnapshot(ctx, rbdSnap, cr)
-			if errDefer != nil {
-				klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), errDefer)
-				err = fmt.Errorf("snapshot created but failed to delete snapshot due to"+
-					" other failures: %v", err)
+			if _, ok := err.(ErrFlattenInProgress); !ok {
+				// cleanup clone and snapshot
+				errCleanUp := cleanUpSnapshot(ctx, cloneRbd, rbdSnap, cloneRbd, cr)
+				if errCleanUp != nil {
+					klog.Errorf(util.Log(ctx, "failed to cleanup snapshot and clone: %v"), errCleanUp)
+				}
 			}
-			err = status.Error(codes.Internal, err.Error())
 		}
 	}()
-	err = protectSnapshot(ctx, rbdSnap, cr)
+
+	err = createSnapshot(ctx, cloneRbd, rbdSnap)
```

Comment:
I am slightly confused by this additional snapshot creation here. Is this the single RBD snapshot from which we will create future clones? As commented [here](https://github.com/ceph/ceph-csi/pull/1160#issuecomment-642069930)?

If so, we again do not seem to be checking if this snapshot is created when we call `checkSnapCloneExists`, as a result we would fail in the future to use this snapshot once it is marked ready?